internal/tjson/schema_test.go

```diff
@@ -67,3 +67,147 @@ func TestSchemaMarshalUnmarshal(t *testing.T) {
+	cInt64Schema := Schema{
+		Type:   Integer,
+		Format: Int64,
+	}
+	cIntEmptySchema := Schema{
+		Type:   Integer,
+		Format: EmptyFormat,
+	}
+	cDoubleSchema := Schema{
+		Type:   Number,
+		Format: Double,
+	}
+	cDoubleEmptySchema := Schema{
+		Type:   Number,
+		Format: EmptyFormat,
+	}
+	cObjectSchema := Schema{
+		Type: Object,
+		Properties: map[string]*Schema{
+			"a":  stringSchema,
+			"42": &cIntEmptySchema,
+		},
+	}
+	cObjectSchemaEqual := Schema{
+		Type: Object,
+		Properties: map[string]*Schema{
+			"42": &cIntEmptySchema,
+			"a":  stringSchema,
+		},
+	}
+	cObjectSchemaNotEqual := Schema{
+		Type: Object,
+		Properties: map[string]*Schema{
+			"42": &cIntEmptySchema,
+			"a":  boolSchema,
+		},
+	}
+	cArrayDoubleSchema := Schema{
+		Type:  Array,
+		Items: &cDoubleSchema,
+	}
+	cArrayDoubleEmptySchema := Schema{
+		Type:  Array,
+		Items: &cDoubleEmptySchema,
+	}
+	cArrayObjectsSchema := Schema{
+		Type:  Array,
+		Items: &cObjectSchema,
+	}
+	cArrayObjectsSchemaEqual := Schema{
+		Type:  Array,
+		Items: &cObjectSchemaEqual,
+	}
+	cArrayObjectsSchemaNotEqual := Schema{
+		Type:  Array,
+		Items: &cObjectSchemaNotEqual,
+	}
+
+	for _, tc := range []struct {
+		name     string
+		s        *Schema
+		other    *Schema
+		expected bool
+	}{{
+		name:     "StringString",
+		s:        stringSchema,
+		other:    stringSchema,
+		expected: true,
+	}, {
+		name:     "StringNumber",
+		s:        stringSchema,
+		other:    doubleSchema,
+		expected: false,
+	}, {
+		name:     "NumberString",
+		s:        doubleSchema,
+		other:    stringSchema,
+		expected: false,
+	}, {
+		name:     "EmptyInt64",
+		s:        &cIntEmptySchema,
+		other:    &cInt64Schema,
+		expected: true,
+	}, {
+		name:     "Int64Empty",
+		s:        &cInt64Schema,
+		other:    &cIntEmptySchema,
+		expected: true,
+	}, {
+		name:     "Int64Int32",
+		s:        &cInt64Schema,
+		other:    int32Schema,
+		expected: false,
+	}, {
+		name:     "EmptyInt32",
+		s:        &cIntEmptySchema,
+		other:    int32Schema,
+		expected: false,
+	}, {
+		name:     "DoubleEmpty",
+		s:        &cDoubleSchema,
+		other:    &cDoubleEmptySchema,
+		expected: true,
+	}, {
+		name:     "ObjectsEqual",
+		s:        &cObjectSchema,
+		other:    &cObjectSchemaEqual,
+		expected: true,
+	}, {
+		name:     "ObjectsNotEqual",
+		s:        &cObjectSchemaEqual,
+		other:    &cObjectSchemaNotEqual,
+		expected: false,
+	}, {
+		name:     "ArrayDouble",
+		s:        &cArrayDoubleSchema,
+		other:    &cArrayDoubleEmptySchema,
+		expected: true,
+	}, {
+		name:     "ArrayObjects",
+		s:        &cArrayObjectsSchema,
+		other:    &cArrayObjectsSchemaEqual,
+		expected: true,
+	}, {
+		name:     "ArrayObjectsNotEqual",
+		s:        &cArrayObjectsSchemaNotEqual,
+		other:    &cArrayObjectsSchemaEqual,
+		expected: false,
+	}, {
+		name:     "ArrayObjectsDouble",
+		s:        &cArrayObjectsSchema,
+		other:    &cArrayDoubleSchema,
+		expected: false,
+	}} {
+		tc := tc
+		t.Run(tc.name, func(t *testing.T) {
```

Comment:
having `t.Parallel` here caused a data race. I believe it's because schemas are pointers, and some of them are reused.

Reply:
We should fix a data race, not ignore it. Reading schema pointers in parallel should be safe, and we should not write to them

Reply:
Fixed